torch_xla/csrc/tensor.cpp

```diff
@@ -210,39 +218,41 @@ torch::lazy::BackendDataPtr XLATensor::GetXlaData() {
   return data()->handle;
 }
 
-XLATensor::ShardingSpecPtr XLATensor::sharding_spec() const {
-  torch::lazy::Value ir_value = CurrentIrValue();
-  if (ir_value) {
-    XLA_CHECK(ir_value.node != nullptr) << "Tyring to access a null cursor";
-    auto sharding = dynamic_cast<XlaNode*>(ir_value.node.get())->GetSharding();
-    if (sharding == nullptr) {
-      return nullptr;
-    }
-    return std::make_shared<ShardingSpec>(*sharding);
-  }
-  return nullptr;
-}
-
 void XLATensor::SetShardingSpec(const ShardingSpec& sharding) {
   // Existing annotation must be cleared explicitly. We do not clear and
   // overwrite the existing sharding on the user's behalf. This is a no-op if
   // the same sharding already applied.
   if (sharding_spec() == nullptr ||
       !ShardingUtil::EqualShardingSpecs(sharding, *sharding_spec())) {
     TORCH_LAZY_COUNTER("SetShardingSpec", 1);
-    XLA_CHECK(GetIrValue().node != nullptr) << "Tyring to access a null cursor";
+    data()->sharding = std::make_shared<ShardingSpec>(sharding);
     dynamic_cast<XlaNode*>(GetIrValue().node.get())
```

Comment:
what's going to happen if `GetIrValue().node` is nullptr? I assume when there we eventually create IR for this node, we will copy the sharding_spec over?

Reply:
Sharded tensor sharding is attached via mark_sharding and it's accompanied with a tensor IR node, backed by sharded device data.

Reply:
Oh, that's because GetIrValue shoulnd't be empty.

Reply:
Do we still need to keep `dynamic_cast<XlaNode*>(GetIrValue().node.get())->SetSharding(sharding.sharding);` through? given that in `CollectSyncTensors` we will set it anyway

Reply:
Yea, I think we could simplify this by removing it, just like in AssignIrValue. Let me test & veirfy though.

Reply:
We can't seem to skip this part, I suspect there are nodes that don't need to be synced (sync tensors). The `AssignIrValue` can be simplified though because it's used to trim after the initial annotation.